### PR TITLE
Prune unused nested SDFG connectors

### PR DESCRIPTION
**Pull request by definelicht**
When input or output connectors of a nested SDFG are not used anywhere inside, we can prune them. This can cascade through multiple levels of nestings with multiple applications of the transformation.

This is used for SDFGs generated from frontends that need to liberally add connectors. It is useful because it can enable other transformations, such as the loop-to-map transformation, which is prevented from overlapping read and write sets.

### CODE REVIEW

**Comment by definelicht** [comment]
`read_and_write_set` will be merged with the one in `loop-to-map` once it's merged.

File: dace/transformation/dataflow/prune_connectors.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                    if state.out_degree(e.src) <= 1:
+                        # If removing this edge would orphan the entry node,
+                        # replace it with an empty edge
+                        state.remove_edge(e)
+                        e.src.remove_out_connector(e.src_conn)
+                        if e.dst == nsdfg:
+                            dst_conn = None
+                        else:
+                            # Maintain the original destination connector
+                            dst_conn = e.dst_conn
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:94]: No need for this

File: dace/transformation/dataflow/prune_connectors.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        for s in nsdfg.sdfg.states():
+            for n in s.data_nodes():
+                all_data_used.add(n.data)
+
+        for conn in prune_in:
+            for e in state.in_edges(nsdfg):
+                if e.dst_conn == conn:
+                    break
+            else:
+                raise RuntimeError("Connector not found.")
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:77]: I think we just need to have functions called `{in,out}_edges_by_connector` or something. We can use them later here as well (see comment below).

File: dace/transformation/dataflow/prune_connectors.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                        e.src.remove_out_connector(e.src_conn)
+                        if e.dst == nsdfg:
+                            dst_conn = None
+                        else:
+                            # Maintain the original destination connector
+                            dst_conn = e.dst_conn
+                        state.add_nedge(e.src, e.dst, memlet.Memlet())
+                    else:
+                        # Otherwise just burninate
+                        state.remove_edge_and_connectors(e)
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:98]: not necessarily remove the connector. what if there is more than one outgoing edge from that connector? (think jacobi)
would be the other use case for `out_edges_by_connector(...)`

File: dace/transformation/dataflow/prune_connectors.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                if isinstance(e.src, nodes.AccessNode):
+                    state.remove_edge_and_connectors(e)
+                    if state.degree(e.src) == 0:
+                        state.remove_node(e.src)
+                elif isinstance(e.src, nodes.EntryNode):
+                    if state.out_degree(e.src) <= 1:
+                        # If removing this edge would orphan the entry node,
+                        # replace it with an empty edge
+                        state.remove_edge(e)
+                        e.src.remove_out_connector(e.src_conn)
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:89]: what about removing the in-connector of the dst node?

File: dace/transformation/dataflow/prune_connectors.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                        state.add_nedge(e.src, e.dst, memlet.Memlet())
+                    else:
+                        # Otherwise just burninate
+                        state.remove_edge_and_connectors(e)
+                else:
+                    raise TypeError("Unexpected node on path: {}".format(
+                        type(e.src)))
+            if conn in nsdfg.in_connectors:
+                # Actually remove the connector if it hasn't already been
+                del nsdfg.in_connectors[conn]
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:104]: `nsdfg.remove_in_connector()`?

File: dace/transformation/dataflow/prune_connectors.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            if conn in nsdfg.sdfg.arrays and conn not in all_data_used:
+                # If the data is now unused, we can also purge it from the SDFG
+                del nsdfg.sdfg.arrays[conn]
+
+        for conn in prune_out:
+            for e in state.out_edges(nsdfg):
+                if e.src_conn == conn:
+                    break
+            else:
+                raise RuntimeError("Connector not found.")
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:114]: same as prune_in

File: dace/transformation/dataflow/prune_connectors.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                if isinstance(e.dst, nodes.AccessNode):
+                    state.remove_edge_and_connectors(e)
+                    if state.degree(e.dst) == 0:
+                        state.remove_node(e.dst)
+                elif isinstance(e.dst, nodes.ExitNode):
+                    if state.out_degree(e.dst) <= 1:
+                        # If removing this edge would orphan the exit node,
+                        # replace it with an empty edge
+                        state.remove_edge(e)
+                        e.dst.remove_in_connector(e.dst_conn)
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:126]: same

File: dace/transformation/dataflow/prune_connectors.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+                        if e.src == nsdfg:
+                            src_conn = None
+                        else:
+                            # Maintain the original source connector
+                            src_conn = e.src_conn
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:131]: same

File: dace/transformation/dataflow/prune_connectors.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                        state.add_nedge(e.src, e.dst, memlet.Memlet())
+                    else:
+                        # Otherwise just burninate
+                        state.remove_edge_and_connectors(e)
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:135]: same

File: dace/transformation/dataflow/prune_connectors.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                else:
+                    raise TypeError("Unexpected node on path: {}".format(
+                        type(e.src)))
+            if conn in nsdfg.out_connectors:
+                # Actually remove the connector if it hasn't already been
+                del nsdfg.out_connectors[conn]
```
Comment (tbennun) [inline dace/transformation/dataflow/prune_connectors.py:141]: same

File: tests/transformations/prune_connectors_test.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+                                memlet=dace.Memlet("read_unused[0:N, 0:N]"))
+    state_outer.add_memlet_path(isolated_nsdfg,
+                                isolated_exit,
+                                isolated_write,
+                                src_conn="write_unused_isolated",
+                                memlet=dace.Memlet("write_unused[0:N, 0:N]"))
+    isolated_state = isolated_sdfg.add_state("isolated")
+		# Cannot figure out a good way to test that this is actually executed the
+    # right number of times: capturing stdout does not work because the SDFG
+		# is run in a subprocess
```
Comment (tbennun) [inline tests/transformations/prune_connectors_test.py:130]: tabs vs. spaces

File: tests/transformations/prune_connectors_test.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    arr_out = np.empty((n, n), dtype=np.uint16)
+
+    sdfg(read_used=arr_in,
+         read_unused=arr_in,
+         write_used=arr_out,
+         write_unused=arr_out,
+         N=n)
+
+    if not all(arr_out.ravel() == arr_in.ravel() + 1):
+        raise ValueError("Validation failed.")
```
Comment (tbennun) [inline tests/transformations/prune_connectors_test.py:161]: pytest format please

File: tests/transformations/prune_connectors_test.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    arr_in = np.zeros((n, n), dtype=np.uint16)
+    arr_out = np.empty((n, n), dtype=np.uint16)
+
+    sdfg(read_used=arr_in,
+         read_unused=arr_in,
+         write_used=arr_out,
+         write_unused=arr_out,
+         N=n)
+
+    if not all(arr_out.ravel() == arr_in.ravel() + 1):
```
Comment (tbennun) [inline tests/transformations/prune_connectors_test.py:160]: `assert np.allclose(arr_out, arr_in + 1)`

**Review by tbennun — CHANGES_REQUESTED** [review]
Good xform and test! Anticipating some corner cases that should be fixed.

Maybe we just need functions like "safely remove this memlet path and all related connectors".

File: dace/transformation/dataflow/prune_connectors.py (hunk 3 above)
Comment (definelicht) [reply dace/transformation/dataflow/prune_connectors.py]: Treated in the refactored function

File: dace/transformation/dataflow/prune_connectors.py (hunk 4 above)
Comment (definelicht) [reply dace/transformation/dataflow/prune_connectors.py]: Treated in the refactored function

File: dace/transformation/dataflow/prune_connectors.py (hunk 2 above)
Comment (definelicht) [reply dace/transformation/dataflow/prune_connectors.py]: Implemented in 0ed2219b and split into two in c2d24f47

File: dace/sdfg/state.py [hunk 14]
```diff
@@ -216,6 +216,34 @@ def traverse(node):
         # Return node that corresponds to current edge
         return traverse(tree_root)
 
+    def in_edges_by_connector(self, node: nd.Node,
+                              connector: AnyStr) -> MultiConnectorEdge:
```
Comment (tbennun) [inline dace/sdfg/state.py:220]: return type should be iterable of edges

File: dace/sdfg/state.py [hunk 15]
```diff
@@ -216,6 +216,34 @@ def traverse(node):
...
+        """ Returns a generator over edges entering the given connector of the
+            given node.
+            :param node: Destination node of edges.
+            :param connector: Destination connector of edges.
+        """
+        return (e for e in self.in_edges(node) if e.dst_conn == connector)
+
+    def out_edges_by_connector(self, node: nd.Node,
+                               connector: AnyStr) -> MultiConnectorEdge:
```
Comment (tbennun) [inline dace/sdfg/state.py:229]: same

File: dace/sdfg/state.py [hunk 16]
```diff
@@ -216,6 +216,34 @@ def traverse(node):
...
+        """ Returns a generator over edges exiting the given connector of the
+            given node.
+            :param node: Source node of edges.
+            :param connector: Source connector of edges.
+        """
+        return (e for e in self.out_edges(node) if e.src_conn == connector)
+
+    def edges_by_connector(self, node: nd.Node,
+                           connector: AnyStr) -> MultiConnectorEdge:
```
Comment (tbennun) [inline dace/sdfg/state.py:238]: same

File: dace/sdfg/state.py [hunk 17]
```diff
@@ -1548,42 +1576,123 @@ def add_memlet_path(self,
...
+                if (isinstance(edge.src, nd.AccessNode)
+                        and self.degree(edge.src) == 0):
+                    self.remove_node(edge.src)
+                if (isinstance(edge.dst, nd.AccessNode)
+                        and self.degree(edge.dst) == 0):
+                    self.remove_node(edge.dst)
+
+        # DEPRECATED FUNCTIONS
+        ######################################
+        def add_array(self,
```
Comment (tbennun) [inline dace/sdfg/state.py]: aww I really thought you managed to remove this :)

File: dace/sdfg/state.py [hunk 18]
```diff
@@ -1548,42 +1576,123 @@ def add_memlet_path(self,
...
+                    other_incoming_same = True
+                    break
+            else:
+                other_incoming_same = False
+                edge.dst.remove_in_connector(edge.dst_conn)
+
+            if isinstance(edge.src, nd.EntryNode):
+                # If removing this edge orphans the entry node, replace the
+                # edge with an empty edge
+                if not other_outgoing:
```
Comment (tbennun) [inline dace/sdfg/state.py:1634]: the condition here should be `if not other_incoming:`.
Think of the following case:
```
       (A) (B)
      /  Map  \
       |     |
       v     v
     /M1 \  / M2\
```

Let's say you remove the right edge (B->Map->M2), then M2 will be disconnected because Map has outgoing edges.

File: dace/sdfg/state.py [hunk 19]
```diff
@@ -1548,42 +1576,123 @@ def add_memlet_path(self,
...
+                    self.add_nedge(edge.src, edge.dst, mm.Memlet())
+                if other_outgoing_same:
+                    # If other inner memlets use the outer memlet, we have to
+                    # stop the deletion here
+                    break
+
+            if isinstance(edge.dst, nd.ExitNode):
+                # If removing this edge orphans the exit node, replace the
+                # edge with an empty edge
+                if not other_incoming:
```
Comment (tbennun) [inline dace/sdfg/state.py:1644]: same as above

**Review by tbennun — CHANGES_REQUESTED** [review]
Minor algorithmic changes

File: dace/sdfg/state.py (hunk 18 above)
Comment (definelicht) [reply dace/sdfg/state.py:1634]: Shouldn't it actually be `if len(self.edges_between(src, dst)) == 0` in both cases? The important thing is to not break any scopes.

Comment (definelicht) [reply dace/sdfg/state.py]: @tbennun What do you think of 37607217?

**Review by tbennun — APPROVED** [review]
LGTM

File: dace/sdfg/state.py (hunk 18 above)
Comment (tbennun) [reply dace/sdfg/state.py]: Much better. Captures exactly what you want (disconnection)
